beacon-chain/sync/initial-sync/blocks_fetcher_test.go

```diff
@@ -743,6 +745,200 @@ func TestBlocksFetcher_filterPeers(t *testing.T) {
+			},
+			update: func(s *peers.BlockProviderScorer) {
+				s.IncrementProcessedBlocks("abc", batchSize*2)
+				s.IncrementProcessedBlocks("def", batchSize*2)
+				s.IncrementProcessedBlocks("ghi", batchSize*2)
+				s.IncrementProcessedBlocks("jkl", batchSize*2)
+				s.IncrementProcessedBlocks("xyz", batchSize*2)
+			},
+			want: []peer.ID{"ghi", "def", "abc", "xyz", "jkl"},
+		},
+		{
+			name: "multiple peers same capacity different scores",
+			args: args{
+				peers: []weightedPeer{
+					{"abc", 9000},
+					{"def", 9000},
+					{"ghi", 9000},
+					{"jkl", 9000},
+					{"xyz", 9000},
+				},
+				peersPercentage: 0.8,
+				capacityWeight:  0.2,
+			},
+			update: func(s *peers.BlockProviderScorer) {
+				s.IncrementProcessedBlocks("xyz", s.Params().ProcessedBlocksCap)
+				s.IncrementProcessedBlocks("def", s.Params().ProcessedBlocksCap/2)
+				s.IncrementProcessedBlocks("ghi", s.Params().ProcessedBlocksCap/4)
+				s.IncrementProcessedBlocks("abc", s.Params().ProcessedBlocksCap/8)
+				s.IncrementProcessedBlocks("jkl", 0)
+			},
+			want: []peer.ID{"xyz", "def", "ghi", "abc"},
+		},
+		{
+			name: "multiple peers different capacities and scores",
+			args: args{
+				peers: []weightedPeer{
+					{"abc", 6500},
+					{"def", 2500},
+					{"ghi", 1000},
+					{"jkl", 9000},
+					{"xyz", 6500},
+				},
+				peersPercentage: 0.8,
+				capacityWeight:  0.2,
+			},
+			update: func(s *peers.BlockProviderScorer) {
+				// Make sure that score takes priority over capacity.
+				s.IncrementProcessedBlocks("ghi", batchSize*5)
+				s.IncrementProcessedBlocks("def", batchSize*15)
+				// Break tie using capacity as a tie-breaker (abc and ghi have the same score).
+				s.IncrementProcessedBlocks("abc", batchSize*3)
+				s.IncrementProcessedBlocks("xyz", batchSize*3)
+				// Exclude peer (peers percentage is 80%).
+				s.IncrementProcessedBlocks("jkl", batchSize)
+			},
+			want: []peer.ID{"def", "ghi", "abc", "xyz"},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mc, p2p, _ := initializeTestServices(t, []uint64{}, []*peerData{})
+			fetcher := newBlocksFetcher(context.Background(), &blocksFetcherConfig{
+				headFetcher:              mc,
+				p2p:                      p2p,
+				peerFilterCapacityWeight: tt.args.capacityWeight,
+			})
+			// Non-leaking bucket, with initial capacity of 10000.
+			fetcher.rateLimiter = leakybucket.NewCollector(0.000001, 10000, false)
+			peerIDs := make([]peer.ID, 0)
+			for _, pid := range tt.args.peers {
+				peerIDs = append(peerIDs, pid.ID)
+				fetcher.rateLimiter.Add(pid.ID.String(), pid.usedCapacity)
+			}
+			if tt.update != nil {
+				tt.update(fetcher.p2p.Peers().Scorers().BlockProviderScorer())
+			}
+			// Since peer selection is probabilistic (weighted, with high scorers having higher
+			// chance of being selected), we need multiple rounds of filtering to test the order:
+			// over multiple attempts, high scorers should be picked on high posstions more often.
```

Comment:
```suggestion
			// over multiple attempts, top scorers should be picked on high positions more often.
```
maybe change high scorers to top scorers